File: projectEuler/31/31.cpp
#include <iostream>
#include <vector>

int coinSums(int n) {
    std::vector<int> coins = {1, 2, 5, 10, 20, 50, 100, 200};
    std::vector<int> ways(n + 1, 0);
    ways[0] = 1;

    for (int coin : coins) {
        for (int j = coin; j <= n; ++j) {
            ways[j] += ways[j - coin];
        }
    }

    return ways[n];
}

int main() {
    int target = 200; // 2 £ in pence
    std::cout << "Number of ways to make " << target << " pence: " << coinSums(target) << std::endl;
    return 0;
}

File: projectEuler/31/31.py
#Coin Sums
#how many different ways can £2 be made using any number of coins?

def coinSums(n):
    coin=[1,2,5,10,20,50,100,200]
    ways=[0]*(n+1)
    ways[0]=1
    for i in range(len(coin)):
        for j in range(coin[i],n+1):
            ways[j]+=ways[j-coin[i]]
    return ways[n]

print(coinSums(200))